proposals/low-level-struct-improvements.md

```diff
@@ -486,189 +480,120 @@ struct S1
-
-The resulting `Span<T>` instance will have a length equal to the size declared on the `fixed` buffer. The *safe-to-escape* scope of the returned value will be equal to the *safe-to-escape* scope of the container.
-
-For each `fixed` declaration in a type where the element type is `T` the language will generate a corresponding `get` only indexer method whose return type is `ref T`. The indexer will be annotated with the `[ThisRefEscapes]` attribute as the implementation will be returning fields of the declaring type. The accessibility of the member will match the accessibility on the `fixed` field.
+    Child _child;
 
-For example, the signature of the indexer for `CharBuffer.Data` will be the following:
-
-```cs
-[ThisRefEscapes]
-internal ref char <>DataIndexer(int index) => ...;
+    // In this case the ref-safe-to-escape of `_child` is to the calling method because that is 
+    // the value of `this` and fields derive it from their receiver. From there method invocation 
+    // rules take over 
+    public ref int Value => ref _child.Value;
+}
 ```
 
-If the provided index is outside the declared bounds of the `fixed` array then an `IndexOutOfRangeException` will be thrown. In the case a constant value is provided then it will be replaced with a direct reference to the appropriate element. Unless the constant is outside the declared bounds in which case a compile time error would occur.
+This require the following changes to the span safety document: 
 
-There will also be a named accessor generated for each `fixed` buffer that provides by value `get` and `set` operations. Having this means that `fixed` buffers will more closely resemble existing array semantics by having a `ref` accessor as well as byval `get` and `set` operations. This means compilers will have the same flexibility when emitting code consuming `fixed` buffers as they do when consuming arrays. This should be operations like `await` over `fixed` buffers easier to emit. 
+- The method invocation rules will include the *ref-safe-to-escape* value of the receiver when it is marked by `[RefThisEscapes]`
+- The parameter lifetime rules will change to note that the *ref-safe-to-escape* of `this` is dependent on `[RefThisEscapes]`
 
-This also has the added benefit that it will make `fixed` buffers easier to consume from other languages. Named indexers is a feature that has existed since the 1.0 release of .NET. Even languages which cannot directly emit a named indexer can generally consume them (C# is actually a good example of this).
+Detailed Notes:
+- An instance method or property annotated with `[RefThisEscapes]` has *ref-safe-to-escape* of `this` set to the *calling method*
+- A `struct` annotated with `[RefThisEscapes]` has the same effect of annotating every instance method and property with `[RefThisEscapes]`
```

Comment:
I think it would be helpful for an example that demonstrates why one may want to annotate a single method/property as opposed to the entire type.

Reply:
"Anonymous structs/unions" used in interop.

There are many cases where `FieldOffset` is not viable because some other member in the struct is variable sized (`nint`, `nuint`, a `pointer`, etc). In those cases, the only solution is to have something like:
```csharp
public unsafe partial struct MyStruct
{
    public _Anonymous_e__Union Anonymous;
    
    public unsafe partial struct _Anonymous_e__Union
    {
        public float m11;
    }
} 
```

In such a scenario, there are a lot of "usability" issues with forcing users to do `value.Anonymous.m11` rather than just allowing `value.m11`. The "simplest" solution here is to expose a `ref property`:
```csharp
public unsafe partial struct MyStruct
{
    public _Anonymous_e__Union Anonymous;
    
    public ref float m11 => ref MemoryMarshal.GetReference(MemoryMarshal.CreateSpan(ref Anonymous.m11, 1));
    
    public unsafe partial struct _Anonymous_e__Union
    {
        public float m11;
    }
} 
```

However, this is unsafe and doesn't properly track the lifetime of the `ref`. With the introduction of `RefThisEscapes`, this becomes:
```csharp
public unsafe partial struct MyStruct
{
    public _Anonymous_e__Union Anonymous;
    
    [RefThisEscapes]
    public ref float m11 => ref Anonymous.m11;
    
    public unsafe partial struct _Anonymous_e__Union
    {
        public float m11;
    }
} 
```

You wouldn't want to annotate the entire type, because things like `operator ==(in MyStruct left, in MyStruct right)` take a ref, but don't return a reference to the `this` field.

Reply:
Thanks. Great example.